### ISSUE TITLE
Allow import blocks to access resources that are created as part of the plan

### ISSUE THREAD

**Issue by abentley-ssimwave** [opentofu/opentofu#1677]
### OpenTofu Version

```shell
1.7
```

### Use Cases

This would be useful for implicitly-created resources.

In AWS, creating a VPC also creates a default ACL, with rules.  These rules cannot be managed by a single `apply` they "already exist" when we attempt to configure them, yet we cannot import them because they don't exist at `plan` time.

### Attempted Solutions

   "import": {                                                                                         
      "id": "${aws_vpc.internal-products.default_network_acl_id}:ingress:100",                         
      "to": "aws_network_acl_rule.ingress_100"                                                         
   },

### Proposal

Import blocks could support an optional argument "precondition", which could be used to defer performing the import until the a particular resource has been fully created.

**Comment by RLRabinowitz** [comment]
Hey @abentley-ssimwave , this behaviour has to do with how the AWS provider has a unique scenario here, managing a default ACL as well as a VPC. The AWS provider also has the `aws_default_network_acl` resource, which allows you to do what you are currently attempting to do, without need for doing any import

Does that your use-case?

**Comment by abentley-ssimwave** [comment]
I disagree that this is a unique scenario.  In GCP, creating a `google_container_cluster` causes a `google_container_node_pool` to be created.  I'm sure there are other similar scenarios.

`aws_default_network_acl` does provide a resource-specific solution, for me (and there's a resource-specific solution for google_container_node_pool also).  However, I think that a general purpose facility would make sense.

Also when changes have been made and you run `plan`, `aws_default_network_acl`'s output does not clearly indicate what has changed, while `aws_network_acl_rule` does.  Due to the way it diffs, `aws_default_network_acl` may show identical versions of a rule being added and deleted, a failing it has in common with `aws_network_acl`.

**Comment by RLRabinowitz** [comment]
OK I see your point here. Up until now, dealing with cases of implicitly created resources in OpenTofu has always been provider and resource specific.

Allowing an import's `id` to be known after apply is something that requires a large scale change in how OpenTofu works. Currently in OpenTofu all plans always show the exact changes to be made in a resource, with the exception of specific fields being known after apply. If the ID is not known at plan time, then you could not tell in the plan which fields of the resource would be updated at apply-time

I'm marking this issue as "needs community input", as we'd wish to hear more input from the community for this change, and also "needs RFC", as there are some things to hash out here

**Comment by JonnyHaystack** [comment]
I have a similar problem, where I need to import config resources from a service that isn't available until a VM has been created and booted up. The resources I'm trying to import to have a dependency on the VM existing, but the import block implementation doesn't seem to be clever enough to have awareness of this.

In my mind an ideal solution would be for it to both:
- Check for dependencies of the target resource (in the `to` field) - this would satisfy my current use case, which I would imagine to be pretty common
- *and* allow an explicit `depends_on` parameter to the import block to specify any additional dependencies that cannot be inferred - this would satisfy @abentley-ssimwave's use case

I don't know the details of any technical barriers to achieving this, but this is what would seem like the most user-friendly interface to me.

**Comment by skyzyx** [comment]
In my previous use-cases, I've leveraged _code generation_ to solve this problem. Maybe it's not a perfect long-term solution, but it may help you get unblocked in the interim.

* Stage 1: Run your code, and specify `output` blocks for the data that matters.

* Stage 2: Pipe the output of `tofu show` to a script/program that can read those IDs and generate the `import` blocks on your behalf. Sometimes I have to fill-in-the-blanks leveraging the AWS SDK, but that's relatively low effort.

Disclaimer: I worked on the AWS SDK team from 2010–2014, so I have a deep understanding of how the SDKs work.

**Comment by abentley-ssimwave** [comment]
I do already have work-arounds, like aws_default_network_acl, and doing things in stages is a general solution.  I just feel like the ideal is that tofu can generate infrastructure from scratch with a single command, not in multiple stages.

**Comment by apparentlymart** [comment]
This request feels conceptually similar to me to how OpenTofu currently treats `data` blocks: if their configuration includes any unknown values and/or they depend on any managed resources that have changes already pending in the current plan then OpenTofu defers asking the provider to read the data until the apply step, which then makes sure that the data source configuration is complete and final before fetching it.

However, the workflow does get a bit tricky in that case: for `data` blocks OpenTofu can use the partially-unknown configuration as a placeholder for planning any downstream resources that depend on the data source, but an `import` block does not provide any configuration for the resource, and even when it's accompanied by a `resource` block that's being imported into there is no guarantee that what is being imported will actually match the configuration; today that's handled by running the import during the planning phase and then having OpenTofu propose an update/replace change to "correct" the imported object to match the configuration, but it would not be acceptable to just introduce an update/replace change discovered after the import during the apply phase because the operator would have no opportunity to review and approve that change.

Therefore I think while this request does seem valid in principle, it raises some significant technical design challenges that need to be addressed before it could be implemented, including but probably not limited to.

- If an import were deferred to the apply phase, how should OpenTofu resolve downstream resources that refer to the resource instance that's being imported into while planning actions for those downstream resources?
- If OpenTofu performs an import during the apply phase and then finds that the imported object doesn't match the desired state of the resource instance it was imported into, what should happen?

---

A possible variation would be for OpenTofu to notice that there's an `import` for a particular resource instance that can't be resolved during the current planning phase due to dependencies, and in that case to behave in a similar way to if that resource instance had been used with the `-exclude` argument: don't plan to do anything for it _at all_, and then have the operator run another plan/apply round to complete the import and thus reach full convergence.

That would essentially be an import-specific incarnation of what I'd proposed _long, long_ ago in https://github.com/hashicorp/terraform/issues/4149: still requires performing the work over two plan/apply rounds, but at least OpenTofu detects the need for it automatically rather than requiring the operator to add a `-target`/`-exclude` argument manually.

Still far easier said than done, but at least it avoids all of the tricky questions that arise from trying to run the import step during the apply phase.
